### ISSUE TITLE
[hdpowerview] PowerView Gen 3 support

### ISSUE THREAD

**Issue by jlaur** [openhab/openhab-addons#12678]
This issue is created as a placeholder for gathering information about the new PowerView Gen 3 and discussing how to support it. I have requested API documentation from Hunter Douglas. They have responded: "This information has not been released to the general public yet, we are working in beta with selected groups as needed."

What we know so far:
- Proprietary RF communication will be replaced by standards-based Bluetooth Low Energy (BLE) protocol.
- The new protocol offers instant two-way synchronization and improved reliability and range.
- A new PowerView Gen 3 Gateway will be introduced, probably similar to PowerView Gen 2 Hub, but communicating over BLE with shades and accessories.
- A new PowerView Pebble Remote will be introduced, communicating over BLE.

Open questions:
- Will the PowerView Gen 3 Gateway support the proprietary RF communication for backwards compatibility with legacy shades and accessories? Answer: No, two hubs/gateways are needed when having a mix of Gen 1/2 and Gen 3 shades.
- Will the PowerView Gen 3 Gateway remain open for integrators like the PowerView Gen 2 Hub? Answer: Yes.
- Will the PowerView Gen 3 Gateway implement a new API and/or support the existing API? Answer: There is a new API - see issue comments.
- Will the PowerView Gen 3 Gateway support the Matter standard at some point?

Sources:
- https://restechtoday.com/hunter-douglas-debuts-powerview-gen-3-automation/
- https://www.cepro.com/control/motorized-shades/hunter-douglas-powerview-gen-3-series/

**Comment by jlaur** [comment]
@andrewfg, @arroyoj - FYI.

**Comment by kingy444** [comment]
The fact they havent released the api is not a great sign for any backwards compatability

**Comment by andrewfg** [comment]
> The fact they havent released the api is not a great sign for any backwards compatability

Or it might mean that the API did not change :)

**Comment by andrewfg** [comment]
The current (older) API uses plain text over HTTP with no encryption and no authentication. So it is pretty archaic. Therefore my guess is that they will have moved to HTTPS and some form of authentication (perhaps password based or token based).

{ Note: I had the exact same experience in the last month with the Heatmiser Neohub where the manufacturer migrated from plain text HTTP to a token based web socket connection instead https://github.com/openhab/openhab-addons/pull/12915#issuecomment-1152413274 and in this case, they kept support for the legacy plain text HTTP but now it requires the user to explicitly switch it on in the App }

**Comment by kingy444** [comment]
Yea that makes sense - would be good if they added it via firmware to the old hubs too tbh - just make it optional

**Comment by andrewfg** [comment]
Yeah. In the Heatmiser App, it is as simple as this..

![image](https://user-images.githubusercontent.com/893994/174585048-858540e6-0e12-462b-bb0d-c862b29ceefc.png)

**Comment by andrewfg** [comment]
> would be good if they added it via firmware to the old hubs

PS in my App there is an option 'Receive Early Access Updates' (Opt in to receive early access to firmware updates before official release) ... however I am extremely nervous about whether or not to enable it ...

**Comment by jlaur** [comment]
> PS in my App there is an option 'Receive Early Access Updates' (Opt in to receive early access to firmware updates before official release) ... however I am extremely nervous about whether or not to enable it ...

IMHO you should try it. :laughing:

**Comment by jlaur** [comment]
Btw, in the linked thread a port scan was made, and port 443 wasn't open.

**Comment by andrewfg** [comment]
> you should try it

LOL

**Comment by andrewfg** [comment]
@jlaur someone made the following post on the HA forum; it's not definitive, but the guy sounds like he knows something..

https://community.home-assistant.io/t/hunter-douglas-powerview-gen-3-integration/424836/17?u=andrewfg

**Comment by andrewfg** [comment]
@jlaur I found the following link, which implies that some commercial companies have developed integrations for Gen3 hubs. The web page has an email link, and I have already emailed them asking for information about the Gen 3 API. I will keep you informed if I get any response.

https://controlconcepts.net/manufacturers-modules-and-drivers/hunter-douglas/

**Comment by andrewfg** [comment]
@jlaur the response was as follows..

> Do you have access to a Gen 3 Gateway and shades? The Gen 3 API documentation lives on the Gateway itself using Swagger. Swagger also allows for easy testing of the REST API, showing command structure and results.
>
> By default, Swagger is disabled on Gateways as it does utilize system resources and is only needed during integration development. http://powerview-g3.local/gateway/swagger?enable=true will enable Swagger. Note that Swagger must be re-enabled after each Gateway power cycle.

**Comment by vves** [comment]
@andrewfg @jlaur Hi, I wanted to introduce myself. I am the product architect for Gen3 can provide some unofficial guidance and support for this integration.  Cheers and Code On!  @jlaur is correct that the starting point for the integration is the Gateway Swagger docs listed above.

**Comment by jlaur** [comment]
> @andrewfg @jlaur Hi, I wanted to introduce myself. I am the product architect for Gen3 can provide some unofficial guidance and support for this integration. Cheers and Code On! @jlaur is correct that the starting point for the integration is the Gateway Swagger docs listed above.

Thanks for stepping in and offering support, @vves. For supporting the Gen 3 gateway we currently lack documentation and/or users (in order to get hands on), since we are currently not in procession of the new gateway. I recently added new TDBU shades to my setup, so I now have 11 shades on a Gen 2 Hub, so for me personally I'm unfortunately nowhere near a Gen 3 upgrade.

**Comment by andrewfg** [comment]
@vves I would also like to thank you for your support; perhaps the best is if you can send us the swagger data (presumably a yaml file?); however alternatively perhaps you can expose one of your hubs via a port forwarding, or vpn connection, so we can talk to it ourselves?

**Comment by kingy444** [comment]
Thanks for reaching out @vves 

I do Dev work on the HomeAssistant side and have been working with @andrewfg closely over the last couple of months (great to see collaboration  between ‘competing’ platforms)

As Andrew suggested if we could get access to the api doco in some alternative way that would be great. 

While offline development without a hub isn’t great (blindly hoping your code logic works) our only other option is to hope another ‘developer’ gets a gen3 hub installed  - I couldn’t afford to spring for new blinds just to help the community out 😂 

Again really appreciate the response - I’d personally tried to reach out to Hunter Douglas and had no response.

Not sure if you have seen some of our other conversations but one thing that would be insanely helpful is the definition of your shade capability and types. Everything we currently have is community driven and relies solely on someone providing their json output to us.

Even just a listing of shades and their capability type would be great for aesthetics but really would be great if some of our ‘guesses’ on capability were substantiated.

**Comment by vves** [comment]
My preference is to use the Gateway Swagger endpoint as the HD dev team is still making minor tweaks to the API.  That said, attached is the current version in a static format so that you guys can start rolling
[gateway-swagger-1.10.0.zip](https://github.com/openhab/openhab-addons/files/9417942/gateway-swagger-1.10.0.zip)
.

**Comment by andrewfg** [comment]
@vves many thanks for your support; much appreciated.

**Comment by andrewfg** [comment]
@jlaur as "homework" for this integration, you may want to take a look at the tado binding which uses Swagger CodeGen to parse the Swagger 'yaml' file and auto- generate the respective Java classes for the DTOs and the web service.

**Comment by andrewfg** [comment]
FWIW a Home Assist user kindly did a screen capture from their Gen 3 hub (see below). It is interesting, but I am still hoping that @vves can post us the yaml file :)

![powerview-gen3-swagger](https://user-images.githubusercontent.com/893994/186924144-a0ad9b2c-469e-40eb-9589-5cea7b753a94.jpeg)

@jlaur of immediate interest to me are the following..
- The urls in Gen 2 hubs like '/api/shades' seem to now be '/homes/shades' instead.
- There seem to be ten capabilities rather than nine ;)
- .. more no doubt to come ..

**Comment by vves** [comment]
@andrewfg Even better - I am putting a 'Getting Started with Integrations' document together.  A deep integration only needs around 10 total calls.  The rest of the API is for internal Hunter Douglas tools and will end up being redacted in the future.

**Comment by andrewfg** [comment]
> putting a 'Getting Started with Integrations' document together

@vves please don't feel that you need to make that extra effort on account of us. We already have full integration for Gen 1 and Gen2 hubs so (dare I say it) we are beyond the 'getting started' phase. All we need from you is the extra information concerning how Gen3 hubs differ from Gen 1/2 ones..

>  A deep integration only needs around 10 total calls

Yes. In fact our integration for the Gen 1/2 hubs goes a bit deeper than ten. We use seventeen calls as you can see from the [HDPowerViewWebTargets.java](https://github.com/openhab/openhab-addons/blob/main/bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/HDPowerViewWebTargets.java) code. Albeit some of these are GET/PUT variations on the same uri.

**Comment by vves** [comment]
@andrewfg there are some significant changes in Gen3 as all shade information is now live and event driven instead of requiring polling for shade and scene state.  It's these changes that I want to highlight and provide best practices for all integrations - not just openHAB.  Coming soon!

**Comment by andrewfg** [comment]
> all shade information is now .. event driven

Ah yes. Had you not mentioned it, I would most probably have overlooked it. ;)

![image](https://user-images.githubusercontent.com/893994/186956303-be7b2e81-bc9f-4a0f-89e4-d688541b9029.png)

**Comment by andrewfg** [comment]
Some data from /home/shades :)
```
[{"id":37,"type":9,"name":"TGVmdA==","ptName":"Left","motion":null,"capabilities":7,"powerType":2,"batteryStatus":3,"roomId":1,"firmware":{"revision":3,"subRevision":0,"build":359},"positions":{"primary":1,"secondary":0,"tilt":0,"velocity":9998},"signalStrength":-55,"bleName":"DUE:1D9E"},{"id":9,"type":9,"name":"Q2VudGVy","ptName":"Center","motion":null,"capabilities":7,"powerType":2,"batteryStatus":3,"roomId":1,"firmware":{"revision":3,"subRevision":0,"build":359},"positions":{"primary":0,"secondary":0,"tilt":0,"velocity":9998},"signalStrength":-61,"bleName":"DUE:8C67"},{"id":13,"type":9,"name":"UmlnaHQ=","ptName":"Right","motion":null,"capabilities":7,"powerType":2,"batteryStatus":3,"roomId":1,"firmware":{"revision":3,"subRevision":0,"build":359},"positions":{"primary":1,"secondary":0,"tilt":0,"velocity":9998},"signalStrength":-66,"bleName":"DUE:C28C"}]
```’

**Comment by andrewfg** [comment]
@jlaur I made some preliminary experiments on implementing the gen 3 api on my branch below..

Basically I started creating two abstract classes `ShadeData` and `ShadePosition` with V1 and V3 implementations `ShadeDataV1`, `ShadeDataV3`, `ShadePositionV1`, `ShadePositionV3`. This means the binding code that depends on `ShadeData` and `ShadePosition` can essentially remain unchanged..

Next step is to also do the same thing with the WebTargets class..

https://github.com/andrewfg/openhab-addons/tree/hdpowerview-generation3/bundles/org.openhab.binding.hdpowerview

**Comment by jlaur** [comment]
@andrewfg - great that you are getting started. I didn't have time since our meeting to think much about the approach, except some initial thoughts.

Collaboration:
- If you are interested in discussing the design and approach to integrating Gen 3 into the existing binding, let's agree on some communication channel/format? I'm fine with this issue for starters, I guess we can extend if needed.
- Let's also sync expectations with each other and Wes about our timeline and effort we can put into this. Perhaps next meeting or PM.
- Pull request strategy: Some refactoring and preparations could be done before actually adding Gen 3 support, or at least it could be done in parallel and frequently rebased. A good strategy could help keeping the initial Gen 3 PR smaller and at the same time ease the maintenance and other work happening in the meantime.
- And of course practical coordination like who will do what.

Technical - as said, not much time for considerations yet:
- **Namespace/structure:** I would suggest to use dedicated subfolders for Gen 3 to avoid naming each individual class something like V3 or V1 (which is actually Gen1+Gen2 as they are almost identical).
- **Naming:** Perhaps we could use "Gen3" as a term. Not sure what to call Gen 1 and Gen 2 combined though. "Legacy" might be a bit ahead of time. :-)
- **Bridge:** What would you think about creating a new bridge type? Name is now "Gateway", not "Hub". Also it doesn't seem to have many similarities with Gen1/2, except hostname. Gen 3 might also have some kind of configuration for SSE (not sure yet).
- **Things:** So far I don't see any need to model things differently, so probably everything can be reused. Only the implementation would differ. What do you think?
- **Controlling multiple shades:** Not sure what is possible - do you know if it is somehow possible to implement/override/intercept `handleCommand` for a group? This could make it possible in the semantic model to iterate through children and create a single request for all shades in the group when using controls on the group.
- Much more to come, but let's start creating ideas. :-)

**Comment by andrewfg** [comment]
Hi @jlaur some quick responses below to your points above..

> agree on some communication channel/format? I'm fine with this issue for starters

Yes, I think this issue is a good place to start. Of course I would be happy to do Zoom calls or whatever. But I think actually writing things down makes it clearer for everybody.

> sync expectations with each other and Wes about our timeline and effort we can put into this.
> Perhaps next meeting or PM.

I think we told Wes that it won't be a fast project. I would suggest that we get back to him in about one month from now.

> Pull request strategy

Yes I agree with the approach of making several small PR's. I created #13339 more as a way of sharing my code ideas with you rather than as an actual PR. So please consider #13339 more as a proof of concept rather than a real PR..

> practical coordination like who will do what.

Yes this makes sense. Perhaps the easiest would be for me to grant you access to my repo (or vice versa) so that both of us can push commits for different parts to the same place?

> Namespace/structure: 

Agreed. To be specific I suggest that for each package where there is a Gen 1 and a Gen 3 alternative, we should have a sub package -- for example as follows. PS I am not too worried about the actual nomenclature (e.g. Gen1/2 vs Gen 3) but to make it as simple as possible I just used the suffixes V1 (aka V1andV2) and V3 .. although we can change that at any time..

- org.openhab.binding.hdpowerview.internal.api.responses
- org.openhab.binding.hdpowerview.internal.api.responses.v1
- org.openhab.binding.hdpowerview.internal.api.responses.v3

> Bridge

I don't think this is necessary. My approach is to make all of the "OH facing" classes agnostic concerning the hub version. Therefore they consume base/abstract classes like `ShadeData`, `ShadePosition`. And the "hub facing" classes `ShadeDataV1`, `ShadePositionV1` (resp. `ShadeDataV3`, `ShadePositionV3`) would have version specific method implementations of the abstract methods of `ShadeData`, `ShadePosition`.

PS apropos the SSE eventing: yes it is true that the Bridge will need to implement a method that is called from the respective WebTarget class. And this method would simply not be used when talking to a V1 hub.

> Things

Ditto as above. The Thing would consume the abstract classes only. 

> Controlling multiple shades

I have no thoughts on this. But I would definitely put it in priority 2

---

Other issues: there are some cases where Gen 3 has less functionality that Gen 1/2 (see below) andd we need to consider how to handle those... 
- Automations: in gen 3 automation is read only, with no API command to enable/disable; so we need to think how to handle this.
- Ditto repeaters..
- Ditto scene collections..
- Ditto battery status (some aspect thereof)..
- Ditto some firmware data..

**Comment by andrewfg** [comment]
Regarding the PRs my suggestion would be the following..
1. Preparation 
- split the current DTO and WebTargets classes into a) a base abstract class, plus b) the current implementation class thereof
- move the current implementation classes into V1 sub folder
- refactor bridge & thing handlers to use the base class; and conditionally do stuff (e.g. battery voltage) that is V1 specific
- refactor JUnit tests (and e.g. builder classes) so that build succeeds
- this version should run perfectly (i.e. not break) V1 hubs

2. Create V3 implementation classes 
- create the V3 implementations of the abstract base classes mentioned above in the V3 sub folder
- refactor bridge & thing handlers to conditionally do stuff (if any) that is V3 specific
- create JUnit tests for V3
- this version should run perfectly (i.e. not break) V1 hubs
- it should also run in a manual config mode on V3 test environments e.g. via Wesley's VPN

3. Go Live
- for functionality that existed on V1 but not on V3 hide the respective channels
- implement discovery for V1 vs V3 hubs and auto set config params to select V1 or V3 mode
- for manual things, add config params to select V1 or V3 mode
- check V3 functionality with real V3 users
- read me

In my PR, I have pretty much completed 1. above already. And the first two bullets of 2. :)

**Comment by andrewfg** [comment]
^
And also my suggestion about who does what..
- @andrewfg - hub facing code (WebTargets, ShadeData dto, ShadePosition dto, Scene dto, ScheduledEvent dto), plus Read Me
- @jlaur - OH facing code (channels, especially non ShadeData ones since you can test) and OH core related stuff (like multi shade commands)

Does that make sense?

**Comment by andrewfg** [comment]
@jlaur the first Pull Request #13339 is ready for your review. This PR changes the architecture by splitting "hub facing" classes into base abstract classes that define the interfaces, and specific implementation classes for V1. This PR should function exaclty as the existing binding since only the code structure has changed.

**Comment by andrewfg** [comment]
@jlaur I have now added the second PR #13355 which is based on the Generation 3 documentation we had already recieved from the HA user community. It does however still require some further work depending on the clarifications that @vves will be providing.

**Comment by andrewfg** [comment]
@jlaur I made the attached PDF to as an aid to navigate the files in my two PR's ..

[Class Structure.pdf](https://github.com/openhab/openhab-addons/files/9505909/Class.Structure.pdf)

**Comment by jlaur** [comment]
@andrewfg - Gen 1/2 vs. Gen 3 detection IMHO should happen during discovery or manual bridge configuration, so that after this point there is no need for pinging any bridge to discover own identity.

I haven't had much time prioritized for this project lately, so I'll be back with input for some of the other open questions.

**Comment by andrewfg** [comment]
> should happen during discovery or manual bridge configuration

I think choosing during discovery would make sense; actually discovery code is something that I am 'reserving' to do in PR #3.. 

Nevertheless I am **NOT** convinced about your suggestion about **manual** configuration: why ask the user to configure something manually when it can easily be done automatically? Perhaps the compromise solution is a kind of auto-manual approach; if the config param is initially empty, then we ping the hubs and set the config param accordingly? After that then the user can (if they have any reason to do so) force override it.

By the way, apropos config params: as you know there are currently two params 'polling interval' and 'hard refresh interval' which I am using as follows..
- V1 hubs:  Execute poll requests continuously at the 'polling interval', and execute hard refreshes continuously at the 'hard refresh interval' (no change from the current binding)
- V2 hubs:  At start up it executes just one poll to initialize the start state, and after that there is no more polling because all state changes are notified via SSE. However at the hard refresh interval it will still execute a poll in order to resynchronise the state in case any SSE events had been accidentally lost, (and it also checks that the SSE socket is still open, and if not, it re-initializes the SSE connection).

> haven't had much time prioritized for this project lately

No problemo :)

**Comment by jlaur** [comment]
> I think choosing during discovery would make sense; actually discovery code is something that I am 'reserving' to do in PR #3..
> 
> Nevertheless I am **NOT** convinced about your suggestion about **manual** configuration

Not much to discuss. Discovery is only for managed things, so for file-based configuration you need to be able to provide this information manually. That's also why there is properties/configuration interoperability (can dig up documentation and/or examples later). For sure I don't want my hub "pinged" on each initialization just because I'm not using managed things.

**Comment by andrewfg** [comment]
>  I don't want my hub "pinged" on each initialization

I don't understand your resistance. The 'ping' is just the get firmare request. And that IS done every time that you initialise the thing. So in reality we are not doing anything different than what happens already.

**Comment by jlaur** [comment]
> I don't understand your resistance. The 'ping' is just the get firmare request. And that IS done every time that you initialise the thing. So in reality we are not doing anything different than what happens already.

That's a valid point. But if the firmware request is different between Gen 1/2 and Gen 3, then it will still make one unneeded call for Gen 3 (from code above), unless I'm missing something?

**Comment by andrewfg** [comment]
^
Yes. On Gen 1 there is no excessive call. Whereas on Gen 3 there will be one extra call to a uri which the hub will fail (with an HTTP 503 error according to the Swagger draft).

**Comment by jlaur** [comment]
> Yes. On Gen 1 there is no excessive call. Whereas on Gen 3 there will be one extra call to a uri which the hub will fail (with an HTTP 503 error according to the Swagger draft).

So where do you see the problem in storing the hub/gateway type during discovery/manual configuration as opposed to having to rediscover it on each initialization?

**Comment by vves** [comment]
I recommend using mdns (bonjour) to check for Gen3 availability.  The Primary gateway and REST interface - if a Gen3 Gateway is online - will always be found at 'powerview-g3.local'.

**Comment by andrewfg** [comment]
> So where do you see the problem in storing the hub/gateway type during discovery/manual configuration as opposed to having to rediscover it on each initialization?

I simply don't want to force the user to do manual configuration.

**Comment by andrewfg** [comment]
> the problem in storing the hub/gateway type during discovery/manual configuration

@jlaur proposed solution as follows..
1. for auto discovered things we can use mDNS as @vves suggests to set a 'representationProperty' called (say) 'generation' that can have an integer value or 1 or 3 depending on which mDNS target had discovered the thing; and when the thing is instantiated that causes a respective configuration parameter of the same name to have that value.
2. for manual defined things, the configuration parameter 'generation' can be **optional**; if it exists (with a value 1 or 3) then the webTargets are created specifically for that generation; and if it does not exist (int value == 0) then we can use my auto-initialise code as follows..
 
```
    private HDPowerViewWebTargets newWebTargets(String host) throws InstantiationException {
        HDPowerViewWebTargets webTargets = this.webTargets;
        if (webTargets != null) {
            return webTargets;
        }
        HDPowerViewHubConfiguration config = getConfigAs(HDPowerViewHubConfiguration.class);
        int hubGeneration = config.generation;
        switch (hubGeneration) {
            case 0: {
                // optional config.generation parameter is missing; instantiate webTargets automatically
                try {
                    // try communicating via V1 web targets
                    webTargets = new HDPowerViewWebTargetsV1(httpClient, host);
                    webTargets.getFirmwareVersions();
                    hubGeneration = 1;
                    break;
                } catch (HubProcessingException | HubMaintenanceException e) {
                    // fall through
                }
                try {
                    // try communicating via V3 web targets
                    webTargets = new HDPowerViewWebTargetsV3(httpClient, host);
                    webTargets.getFirmwareVersions();
                    hubGeneration = 3;
                    break;
                } catch (HubProcessingException | HubMaintenanceException e) {
                    // fall through
                }
                break;
            }
            case 1: {
                webTargets = new HDPowerViewWebTargetsV1(httpClient, host);
                break;
            }
            case 3: {
                webTargets = new HDPowerViewWebTargetsV3(httpClient, host);
                break;
            }
        }
        if (webTargets != null) {
            this.webTargets = webTargets;
            return webTargets;
        }
        throw new InstantiationException("Unable to instantiate the web targets");
    }
```

**Comment by vves** [comment]
@andrewfg We are on the same page.  First time auto-discovery using both mdns addresses is user friendly as (from our experience) most users are not aware of Gen1/2/3 differences or even what generation they have -  and then cache what you found and configured for future use to reduce your integrations boot time.

**Comment by jlaur** [comment]
> We are on the same page. First time auto-discovery using both mdns addresses is user friendly as (from our experience) most users are not aware of Gen1/2/3 differences or even what generation they have - and then cache what you found and configured for future use to reduce your integrations boot time.

This is exactly what I [proposed](https://github.com/openhab/openhab-addons/issues/12678#issuecomment-1240814999), so now it seems all three of us are on the same page. :-) Perhaps the confusion was from the fact that we have both managed and unmanaged things in openHAB. Managed things can be created from discovery and would then according to my proposal be "born" with the Gen version cached. For unmanaged things (configured in files) manual configuration is needed in order for them to have the same information at hand. Thereby they will also have reduced boot time/initialization phase, which was my point. And on top of that, there will be no differences in the logic after creation, so the implementation will be simpler.

@andrewfg - we can consider the lack of configured "Gen" version configuration as Gen 1/2. That will provide the backwards compatibility and eliminate any need for detecting Gen version for already configured things. We could simply use 0 for Gen 1/2 and 1 for Gen 3, unless there would ever be any need to distinguish between Gen 1 and 2, @vves?

**Comment by andrewfg** [comment]
> consider the lack of configured "Gen" version configuration as Gen 1/2. That will provide the backwards compatibility and eliminate any need for detecting Gen version for already configured things.

Yes. Ok.

> We could simply use 0 for Gen 1/2 and 1 for Gen 3

Not quite. I suggest the following

```
        HDPowerViewHubConfiguration config = getConfigAs(HDPowerViewHubConfiguration.class);
        int hubGeneration = config.generation;
        switch (hubGeneration) {
            case 0: 
            case 1: 
            case 2:
                webTargets = new HDPowerViewWebTargetsV1(httpClient, host);
                break;
            case 3:
                webTargets = new HDPowerViewWebTargetsV3(httpClient, host);
                break;
        }
        if (webTargets != null) {
            this.webTargets = webTargets;
            return webTargets;
        }
    }
```

**Comment by andrewfg** [comment]
>  unless there would ever be any need to distinguish between Gen 1 and 2

@jlaur as far as I recall we recently had an issue (see below) where the functionality needed to be different for Generation 1 than Generation 2. And I think you also had something about scene groups not being supported in Generation 1? So we could consider re-factoring those PR's to handle those Gen 1 / 2 differences based on `config.generation`, rather than however we actually did it in those PR's. .. (??) .. But on the other hand, we might just _'let sleeping dogs lie'_..

https://community.openhab.org/t/hd-powerview-binding-issue-openhab-3-3-0/136921

**Comment by jlaur** [comment]
@andrewfg - the only two differences known so far:
- Gen 1 doesn't provide capabilities, only type.
- Gen 2 redirects `/api/scenecollections` to `/api/sceneCollections` while Gen 1 only supports `/api/scenecollections`.

For both of these differences I'm not sure if it's worth distinguishing **API version**, although we **could** get a small optimization for Gen 2 out of directly using `/api/sceneCollections` to avoid redirection.

But that leads back to what should be persisted in bridge config. We have currently two different API's and three different hardware units (Gen 1, Gen 2 and Gen 3). Especially for the Gen 3 unit, the API is likely to evolve over time, so software version might be significant to look at also. However, since software version can change after creation of bridge, we can probably only reliably detect hardware version at time of discovery. So we might have to dynamically handle **minor** differences in API versions after that, for example based on firmware version detected.

Now, the question is if we should link configuration directly to hardware version, or if we simply need to have incremental API **major** version numbering in order to **abstract** the hardware completely? Gen 4 might be based on the same API as Gen 3, so I'm not sure we'll have benefit of storing "3", and we have no guarantee we'll be able to detect actual hardware version rather than API version. Like "1" and "2" would pretty much refer to the same API version, namely current API.

**Comment by andrewfg** [comment]
^
It sounds like you are discussing naming issues again?

Probably the users are more aware of what hardware version they have. So the primary user decision point is if they have a gen1, gen2, or gen3 hardware. For that reason I think the config param should reflect the hardware too.

The secondary (and for the user hidden) point is the api version that runs on that hardware.

As far as I am concerned the gen1 and gen2 hardware both support the V1 api. Albeit with some minor nuances, as you hilighted above.

And the gen3 hub supports the V3 api.

So to summarise we have three hub hardware versions, (representated by an integer config param), that map to two api versions.

- hardware v 1, 2 (or unknown =0) uses V1 api
- hardware v 3 uses V3 api

**Comment by andrewfg** [comment]
@jlaur some background on my two PRs..
- the first PR is simply refactoring the existing functionality to split classes into a non version specific declaration class, and a (V1) version specific implementation class; there are 24 files involved.
- the second PR adds implementaion classes for V2; it includes the 24 changes above, plus a further 27 for V3, Junit tests, and mDNS discovery; in total 51 files have changed.

So if you are reviewing my code, you may consider whether to bother with looking at PR 1 or whether to jump straight into PR 2. Its up to you..

FYI I myself am already running the full PR 2 code on my V1 operative system.

**Comment by andrewfg** [comment]
@vves just wondering what is the status concerning your revised document? In particular the api call for the binding to register itself with the gateway.
